### ISSUE TITLE
Understanding register_full_backward_hook method

### ISSUE THREAD

**Issue by JunweiZheng93** [pytorch/pytorch#86051]
#### Potential bug:
If I understand correctly, when we call `output.backward()`, the gradient wrt the model input, output and weights will be calculated and saved in `grad_input`, `grad_output` and `param_grad` respectively (check the snippet below). After that, `bp_hook` is called. In this case, the gradient of the model inside and outside bp_hook should be the same. The fact is, they are not the same.

#### Reproducing example:
```python
import torch  # torch version is 1.11.0

def bp_hook(module, grad_input, grad_output):
    param_grad = list(module.parameters())[0].grad
    print(f'gradient of the module inside bp_hook: {param_grad}')

net = torch.nn.Linear(4, 1, bias=False)
net.register_full_backward_hook(bp_hook)
data = torch.ones(1, 4)
output = net(data)
output.backward()
print(f'gradient of the module outside bp_hook: {list(net.parameters())[0].grad}')
```
```python
gradient of the module inside bp_hook: None
gradient of the module outside bp_hook: tensor([[1., 1., 1., 1.]])
```

#### Expected behaviour:
Gradient of the model inside and outside bp_hook should be the same.

### Versions

PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: CentOS Linux release 7.9.2009 (Core) (x86_64)
GCC version: (GCC) 8.3.1 20190311 (Red Hat 8.3.1-3)
Clang version: Could not collect
CMake version: version 2.8.12.2
Libc version: glibc-2.17

Python version: 3.9.12 (main, Jun  1 2022, 11:38:51)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.66.1.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 2080 Ti
GPU 1: NVIDIA GeForce RTX 2080 Ti
GPU 2: NVIDIA GeForce RTX 2080 Ti
GPU 3: NVIDIA GeForce RTX 2080 Ti
GPU 4: NVIDIA GeForce RTX 2080 Ti
GPU 5: NVIDIA GeForce RTX 2080 Ti
GPU 6: NVIDIA GeForce RTX 2080 Ti
GPU 7: NVIDIA GeForce RTX 2080 Ti

Nvidia driver version: 510.47.03
cuDNN version: /usr/lib64/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] torch==1.11.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.3.1               h2bc3f7f_2
[conda] mkl                       2022.0.1           h06a4308_117
[conda] numpy                     1.22.3                   pypi_0    pypi
[conda] pytorch                   1.11.0          py3.9_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by albanD** [comment]
Hi,

The gradient wrt the parameters is not ready during the hook no. Only the passed-in grad_output and grad_input should be accessed from there.
If you want to access the gradients of the parameters, you can attach hook to those directly.

**Comment by JunweiZheng93** [comment]
Thx for the reply. That makes sense. I am now a bit curious why grad_input and grad_output are both ready during the hook but the gradient wrt the parameters is not.

**Comment by JunweiZheng93** [comment]
> Hi,
> 
> 
> 
> The gradient wrt the parameters is not ready during the hook no. Only the passed-in grad_output and grad_input should be accessed from there.
> 
> If you want to access the gradients of the parameters, you can attach hook to those directly.

Thx for the reply. That makes sense. I am now a bit curious why grad_input and grad_output are both ready during the hook but the gradient wrt the parameters is not.

**Comment by albanD** [comment]
Only by design of the hook. Since we aim at providing grad_input and grad_output to the user. These are the only things we ensure are ready.

**Comment by soulitzer** [comment]
In this case, inputs do not require grad so it would not be necessary to compute the grad_inputs, so the full backward hook is actually executed BEFORE the backward for the module is run.

@albanD (for the usual case when the grad_inputs were required) since the backward graph always guaranteed to run in the reverse order of the inputs and accumulate grad nodes are prioritized, should we expect the gradients to be read during the hook? Maybe this is not true about the multithreaded case though, but I'm not sure.

**Comment by albanD** [comment]
>  the backward graph always guaranteed to run in the reverse order

It is not guaranteed as soon as multiple devices are involved and I don't think we want to force ourselves to do that.
